cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -305,76 +327,173 @@ func (s *SegmentFilter) UnmarshalJSON(p []byte) error {
 }
 
 type PaginationOptions struct {
-	Cursor *events.Cursor `json:"cursor,omitempty"`
-	Limit  uint           `json:"limit,omitempty"`
+	Cursor *db.Cursor `json:"cursor,omitempty"`
+	Limit  uint       `json:"limit,omitempty"`
 }
 
 type GetEventsResponse struct {
 	Events       []EventInfo `json:"events"`
 	LatestLedger uint32      `json:"latestLedger"`
 }
 
-type eventScanner interface {
-	Scan(eventRange events.Range, f events.ScanFunction) (uint32, error)
-}
-
 type eventsRPCHandler struct {
-	scanner      eventScanner
+	dbReader     db.EventReader
 	maxLimit     uint
 	defaultLimit uint
+	logger       *log.Entry
+	ledgerReader db.LedgerReader
+}
+
+func combineContractIDs(filters []EventFilter) ([][]byte, error) {
+	contractIDSet := set.NewSet[string](maxFiltersLimit * maxContractIDsLimit)
+
+	for _, filter := range filters {
+		for _, contractID := range filter.ContractIDs {
+			contractIDSet.Add(contractID)
+		}
+	}
+
+	contractIDs := make([][]byte, 0, len(contractIDSet))
+	for contractID := range contractIDSet {
```

Comment:
Wait sorry I'm a little confused: what's the point of building the set actually?

Reply:
Oh duh it's to prevent duplicate contract IDs..

But then you could build the array in the same loop and check against `contractIDSet.Has(...)`, right?

Reply:
oh yeah that is much cleaner actually 👍

Reply:
well both are equally ugly lol, I liked the previous one better but okay with this too.